### PR TITLE
Duplicate packets on start

### PR DESCRIPTION
**Pull request by seanchen1991**
Closes: #2093 

## Description

This PR changes the `handle_packet_cmd` flow so that it initiates the clearing of packets upon receiving a batch of `IbcEvents`. 

The issue seems to have originated as a result of some non-determinism in how the relayer subscribes to chains. The usual expected flow is that upon subscribing to a chain, a `NewBlock` command is received at some height `h`, followed by an `IbcEvents` command also at height `h`. However, sometimes the `NewBlock` command at height `h` would be missed and the relayer would only receive the `IbcEvents` command at height `h`. This command would be processed at height `h`, and then later a subsequent `NewBlock` command at associated with height `h + 1` would arrive, which would trigger a `clear_pending_packets` flow at height `h`, which resulted in a "packet messages are redundant" error since the relayer would attempt to clear the `IbcEvents` at height `h`, but they had already been cleared. 

The key insight was realizing that the relayer needs to clear pending packets regardless of whether the `NewBlock` command arrives before the `IbcEvents` command or vice versa. Regardless of the ordering of these two commands, pending packets at height `h - 1` and before need to be cleared _once_ on startup if the relayer has been configured to do so. 
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [x] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: relayer/src/worker/packet.rs [hunk 1]
```diff
@@ -172,6 +169,9 @@ fn handle_packet_cmd<ChainA: ChainHandle, ChainB: ChainHandle>(
                 clear_interval,
                 height,
             ) {
+                // Decrement the height in order to avoid clearing packets at the same
+                // height at which the packet's events were queried
+                let height = height.decrement().unwrap();
```
Comment (seanchen1991) [inline relayer/src/worker/packet.rs:174]: Is unwrapping here upon decrementing acceptable?

**Comment by plafer** [comment]
Can we add a test here? Seems like the kind of bug that can resurface easily after a refactor

File: relayer/src/worker/packet.rs (hunk 1 above)
Comment (ancazamfir) [reply relayer/src/worker/packet.rs:174]: I think we can do something like:
```
    match height.decrement() {
        Err(e) => Err(LinkError::decrement_height(height, e)),
        Ok(height) => link.a_to_b.schedule_packet_clearing(Some(height)),
    }
```

**Comment by ancazamfir** [comment]
Looked a bit through the logs and what happens is that depending on the event registration, hermes may miss the `NewBlock` for `h` but get an `IbcEvent` for `h`. So now it will process the events at `h` and on `NewBlock` for `h+1` it will clear for `h` (sending duplicate packets).

The fix is to clear packets at `h-1` either on `NewBlock` **or** `IbcEvent` command (with height `h`) from supervisor.

**Comment by ancazamfir** [comment]
You may want to wait for @soareschen's PR https://github.com/informalsystems/ibc-rs/pull/2238 as the changes there will affect this PR and you can make use the refactoring there.

File: relayer/src/worker/packet.rs [hunk 2]
```diff
@@ -115,34 +115,54 @@ fn handle_packet_cmd<ChainA: ChainHandle, ChainB: ChainHandle>(
...
+
+    if *should_clear_on_start {
+        *should_clear_on_start = false;
+        return true;
+    }
+
+    let height = height.unwrap();
+
     clear_interval != 0 && height.revision_height % clear_interval == 0
 }
```
Comment (seanchen1991) [inline relayer/src/worker/packet.rs:167]: In order to reduce duplicate checks of the `should_clear_on_start` flag in the `handle_packet_cmd` function, the boolean was moved into this function, but I'm not terribly keen on this refactor, as it bloats the logical complexity of what was originally a very simple check. I'm also not entirely sure that this logic is entirely sound; it was merely the cleanest way I came up with to implement this.

File: relayer/src/worker/packet.rs [hunk 3]
```diff
@@ -99,50 +99,87 @@ pub fn spawn_packet_cmd_worker<ChainA: ChainHandle, ChainB: ChainHandle>(
...
+    } else {
+        Ok(())
     }
 }
 
-/// Whether or not to clear pending packets at this `step` for the given height.
-/// Packets are cleared if `clear_interval` is not `0` and if we have reached the interval.
-fn should_clear_packets(clear_interval: u64, height: Height) -> bool {
+/// Whether or not to clear pending packets at this `step` for some height.
+/// If the relayer has been configured to clear packets on startup and that has not
```
Comment (ancazamfir) [inline relayer/src/worker/packet.rs:162]:
```suggestion
/// If the relayer has been configured to clear packets on start and that has not
```

File: relayer/src/worker/packet.rs [hunk 4]
```diff
@@ -99,50 +99,87 @@ pub fn spawn_packet_cmd_worker<ChainA: ChainHandle, ChainB: ChainHandle>(
...
+    //     return false;
+    // }
+
+    // if *should_clear_on_start {
+    //     *should_clear_on_start = false;
+    //     return true;
+    // }
+
+    // let height = height.unwrap();
+
```
Comment (ancazamfir) [inline relayer/src/worker/packet.rs:182]: Can you remove the commented code? Also further down.

**Review by ancazamfir — APPROVED** [review]
The changes look good, thanks @seanchen1991! some commented out code needs to be cleaned. I will approve as I am off tomorrow.

**Pull request merged**